File: solutions/capabilities/updating/domain/value-objects/StopList.ts
// The yank list: versions clients must refuse even though a manifest offers them.
//
// Ported from V1 tools/updater/src/stop_list.rs (renamed to kebab-case to match
// the surrounding TypeScript).

export interface YankedEntry {
  version: string;
  reason: string | null;
  yanked_at: string;
}

export interface StopListData {
  yanked: YankedEntry[];
  generated_at: string;
}

export class StopList implements StopListData {
  yanked: YankedEntry[];
  generated_at: string;

  constructor(data: StopListData) {
    this.yanked = data.yanked;
    this.generated_at = data.generated_at;
  }

  static fromJson(json: string): StopList {
    const parsed = JSON.parse(json) as StopListData;
    return new StopList({
      yanked: parsed.yanked ?? [],
      generated_at: parsed.generated_at,
    });
  }

  toJson(): string {
    return JSON.stringify({ yanked: this.yanked, generated_at: this.generated_at }, null, 2);
  }
}

export function isYanked(stopList: StopListData, version: string): YankedEntry | undefined {
  return stopList.yanked.find((entry) => entry.version === version);
}

File: solutions/capabilities/updating/rust/stop_list.rs
use serde::{Deserialize, Serialize};
use anyhow::Result;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct YankedEntry {
    pub version: String,
    pub reason: Option<String>,
    pub yanked_at: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct StopList {
    pub yanked: Vec<YankedEntry>,
    pub generated_at: String,
}

pub fn is_yanked<'a>(stop_list: &'a StopList, version: &str) -> Option<&'a YankedEntry> {
    stop_list
        .yanked
        .iter()
        .find(|entry| entry.version == version)
}

impl StopList {
    pub fn from_json(json: &str) -> Result<Self> {
        Ok(serde_json::from_str(json)?)
    }

    pub fn to_json(&self) -> Result<String> {
        Ok(serde_json::to_string_pretty(self)?)
    }
}
